legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-physical-plan/src/main/java/org/finos/legend/engine/persistence/components/schemaevolution/SchemaEvolution.java

```diff
@@ -138,32 +150,51 @@ private List<Operation> stagingToMainTableColumnMatch(Dataset mainDataset,
                         }
                         // If the datatype is a non-breaking change, we alter the datatype.
                         // We also alter the length if required (pick the maximum length)
-                        else if (sink.capabilities().contains(Capability.EXPLICIT_DATA_TYPE_CONVERSION) && sink.supportsExplicitMapping(matchedMainField.type().dataType(), stagingFieldType.dataType()))
+                        else if (sink.capabilities().contains(Capability.EXPLICIT_DATA_TYPE_CONVERSION)
+                                && sink.supportsExplicitMapping(matchedMainField.type().dataType(), stagingFieldType.dataType()))
                         {
-                            //Modify the column in main table
-                            Field newField = evolveFieldLength(matchedMainField, stagingField);
-                            evolveDataType(newField, matchedMainField, mainDataset, operations, modifiedFields);
+                            if (schemaEvolutionCapabilitySet.contains(SchemaEvolutionCapability.DATA_TYPE_CONVERSION))
+                            {
+                                //Modify the column in main table
+                                Field newField = evolveFieldLength(matchedMainField, stagingField);
+                                evolveDataType(newField, matchedMainField, mainDataset, operations, modifiedFields);
+                            }
+                            else
+                            {
+                                throw new IncompatibleSchemaChangeException(String.format("Explicit data type conversion from \"%s\" to \"%s\" couldn't be performed since user capability does not allow it", matchedMainField.type().dataType(), stagingFieldType.dataType()));
+                            }
                         }
 
                         //Else, it is a breaking change. We throw an exception
                         else
                         {
-                            throw new IncompatibleSchemaChangeException("Breaking schema change from datatype " + matchedMainField.type().dataType() + " to " + stagingFieldType.dataType());
+                            throw new IncompatibleSchemaChangeException(String.format("Breaking schema change from datatype \"%s\" to \"%s\"", matchedMainField.type().dataType(), stagingFieldType.dataType()));
                         }
                     }
+                    //If data types are same, we check if length requires any evolution
                     else
                     {
                         Field newField = evolveFieldLength(stagingField, matchedMainField);
                         evolveDataType(newField, matchedMainField, mainDataset, operations, modifiedFields);
                     }
                 }
+                //If Field types are same, we check to see if nullability needs any evolution
+                else
```

Comment:
Do we need an else here ?
It's possible to have both data type change and nullability change ?